### ISSUE TITLE
Lib: args rest type should always be array of any

### ISSUE THREAD

**Issue by antongolub** [microsoft/TypeScript#36327]
**TypeScript Version:**  3.8.0-dev.20200119

**Search Terms:** rest, ConstructorParameters, ReturnType
**Code**

```ts
/**
 * Obtain the parameters of a constructor function type in a tuple
 */
type ConstructorParameters<T extends new (...args: any) => any> = T extends new (...args: infer P) => any ? P : never;

```

**Expected behavior:**
`...args` rest param should extend `any[]`

**Actual behavior:**
`...args` has `any` type

**Playground Link:** 

**Related Issues:**

**Comment by dragomirtitian** [comment]
Why is this an issue ? `any` is assignable to anything, including to `any[]` and ca be the type of a rest parameter. Is this causing any problems anywhere? I can't think of any problems this would cause.

**Comment by antongolub** [comment]
I experimented with mixin factories and got  signature mismatch of my `IConstructable` type when I unfolded the inheritance chain and tried to check the type through InstanceType<typeof MixinResultClass>

```
type IConstructable<T= {}, A extends any[] = any[]> = IConstructor<T, A>
<T extends new (...args: any) => any>
```

As I understand `any[]` extends `any`, but `any` does not extend `any[]`

**Comment by dragomirtitian** [comment]
@antongolub Maybe post a self contained example reproducing the issue?

**Comment by antongolub** [comment]
Well, once I tried to localize problem point, I've found accidental iface and variable names clash.
```typescript
interface D {
}
// ~200 lines of code
const D = ... 
```
This made a major contribution to my problem.
So this PR improvement does not make any sense to my original goal: verify that custom `IConstructables` corresponds [TS expectations](https://github.com/microsoft/TypeScript/issues/17572). My hypothesis was that I could check `new ()...` method signature vs `T extends new(args: any[])` condition. I was wrong. For example, `new(...args: null)`  in some interface passes the check.

```typescript
type InstanceType<T> = T extends new (...args: any[]) => infer R ? R : any;

interface Foo { foo: string }

interface WeirdConstructable {
  new (...args: never): Foo
}

const instance: InstanceType<WeirdConstructable> = { foo: 1 } // $ExpectError
```
